sonar-kotlin-checks/src/main/java/org/sonarsource/kotlin/checks/UselessNullCheckCheck.kt

```diff
@@ -0,0 +1,194 @@
+            handleNonNullCheck(
+                kfc,
+                unaryExpression.baseExpression!!,
+                unaryExpression.operationReference
+            ) {
+                +"non-null assertion "
+                code("!!")
+            }
+        }
+    }
+
+    override fun visitCallExpression(callExpression: KtCallExpression, kfc: KotlinFileContext) {
+        val resolvedCall = callExpression.getResolvedCall(kfc.bindingContext)
+        if (resolvedCall matches NON_NULL_CHECK_FUNS) {
+            // requireNotNull and checkNotNull have no implementations without parameters. The first parameter is always the value to check.
+            handleNonNullCheck(
+                kfc,
+                callExpression.getArgumentByParameterIndex(0, kfc.bindingContext).first().getArgumentExpression()!!,
+                callExpression
+            ) {
+                +"non-null check "
+                code(resolvedCall!!.resultingDescriptor.name.asString())
+            }
+        }
+    }
+
+    private fun KtBinaryExpression.operandComparedToNull(bc: BindingContext): KtExpression? {
+        val left = left?.predictRuntimeValueExpression(bc) ?: return null
+        val right = right?.predictRuntimeValueExpression(bc) ?: return null
+
+        return when {
+            left.isNull() -> right
+            right.isNull() -> left
+            else -> null
+        }
+    }
+
+    private fun handleNullCheck(
+        kfc: KotlinFileContext,
+        expression: KtExpression,
+        issueLocation: PsiElement,
+        nullCheckTypeForMessage: Message.() -> Unit
+    ) =
+        raiseIssueIfUselessCheck(kfc, expression, issueLocation, true, nullCheckTypeForMessage)
+
+    private fun handleNonNullCheck(
+        kfc: KotlinFileContext,
+        expression: KtExpression,
+        issueLocation: PsiElement,
+        nullCheckTypeForMessage: Message.() -> Unit
+    ) =
+        raiseIssueIfUselessCheck(kfc, expression, issueLocation, false, nullCheckTypeForMessage)
+
+    private fun raiseIssueIfUselessCheck(
+        kfc: KotlinFileContext,
+        expression: KtExpression,
+        issueLocation: PsiElement,
+        comparesToNull: Boolean,
+        nullCheckTypeForMessage: Message.() -> Unit
+    ) {
+        if (kfc.mayBeAffectedByErrorInSemantics()) return
+
+        val (nullCaseResult, nonNullCaseResult) = if (comparesToNull) {
+            "succeeds" to "fails"
+        } else {
+            "fails" to "succeeds"
+        }
+
+        val resolvedExpression = expression.predictRuntimeValueExpression(kfc.bindingContext)
+
+        val result = if (resolvedExpression.isNull()) {
+            nullCaseResult
+        } else if (
+        // We are not using the resolvedExpression on purpose here, as it can cause FPs. See SONARKT-373.
+            expression.isNotNullable(kfc.bindingContext)
+        ) {
+            nonNullCaseResult
+        } else {
+            return
+        }
```

Comment:
This conditional logic is a bit hard to read at first glance, would you reconsider rearranging it? For example:
```suggestion
        val result = if (expression.predictRuntimeValueExpression(kfc.bindingContext).isNull()) {
            if (comparesToNull) "succeeds" else "fails"
        } else if (expression.isNotNullable(kfc.bindingContext)) { // Not using the resolvedExpression as it can cause FPs. See SONARKT-373.
            if (comparesToNull) "fails" else "succeeds"
        } else {
            return
        }
```